File: android/sdui-core/src/main/java/com/nba/sdui/core/data/SduiRepository.kt
package com.nba.sdui.core.data

import android.util.Log
import com.nba.sdui.core.models.generated.SduiModels
import com.nba.sdui.core.models.generated.mapper
import com.nba.sdui.core.request.RequestEnvelopeBuilder
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.logging.HttpLoggingInterceptor
import java.util.UUID
import java.util.concurrent.TimeUnit

/**
 * SDUI Repository - Fetches SDUI responses from the composition service.
 *
 * All screen fetches route through [fetchScreen].  There are NO endpoint-specific
 * helpers — the server path is resolved upstream (by resolveEndpoint or the
 * server's navigation payload) and passed here as-is.
 *
 * Request context (platform, locale, device, experiments) is sent as
 * bracket-notation query parameters per plan-request-transport.md D1.
 * If the query string exceeds 8192 chars, falls back to POST with JSON body.
 */
class SduiRepository(
    private val baseUrl: String,
    private val httpClient: OkHttpClient = sharedHttpClient,
    private val authorizationToken: String? = null
) {
    companion object {
        private const val TAG = "SduiRepository"
        private val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()

        /** Shared OkHttpClient — connection pool & thread pool reused across screens. */
        val sharedHttpClient: OkHttpClient by lazy {
            OkHttpClient.Builder()
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .writeTimeout(30, TimeUnit.SECONDS)
                .addInterceptor(HttpLoggingInterceptor().apply {
                    level = HttpLoggingInterceptor.Level.BASIC
                })
                .build()
        }

        /** Backwards-compat alias for callers that referenced the old name. */
        @Deprecated("Use sharedHttpClient", ReplaceWith("sharedHttpClient"))
        val httpClient: OkHttpClient
            get() = sharedHttpClient
    }

    /**
     * Fetch any SDUI screen by its resolved server path.
     *
     * @param path        Path relative to baseUrl, e.g. "/sdui/game-detail/00423"
     * @param envelope    Request envelope builder with platform/device/experiment context
     * @param userParams  Optional user-supplied filter params (e.g. Form submit
     *                    bindings like `season=2025-26`). Always travel in the
     *                    URL query string regardless of GET vs POST so the
     *                    server reads them through the same `@RequestParam`
     *                    path on either side. They participate in the
     *                    GET/POST length decision.
     * @param traceIdOverride  Optional trace ID to reuse from a parent fetch
     *                    (e.g. parameterized refresh inheriting its screen's
     *                    trace). Falls back to `envelope.generateTraceId()`
     *                    when null.
     */
    suspend fun fetchScreen(
        path: String,
        envelope: RequestEnvelopeBuilder,
        userParams: Map<String, String> = emptyMap(),
        traceIdOverride: String? = null
    ): SduiModels = withContext(Dispatchers.IO) {
        val traceId = traceIdOverride ?: envelope.generateTraceId()
        val envelopeQuery = envelope.buildQueryString()
        val userQuery = encodeUserParams(userParams)
        // Threshold includes both halves so large userParams trigger POST too.
        val combinedLength = envelopeQuery.length + (if (userQuery.isEmpty()) 0 else userQuery.length + 1)

        val request = if (combinedLength > 8192) {
            // POST fallback for oversized envelopes. The envelope travels as
            // a JSON body (same shape as SduiRequestContext) so the server's
            // BracketParamResolver deserializes it identically to the
            // iOS / web POST paths. User params still ride the URL so the
            // server's @RequestParam handler reads them the same way as on GET.
            val url = if (userQuery.isEmpty()) "$baseUrl$path" else "$baseUrl$path?$userQuery"
            Log.d(TAG, "Fetching screen (POST fallback): $url")
            Request.Builder()
                .url(url)
                .header("X-Trace-Id", traceId)
                .header("X-Request-Id", UUID.randomUUID().toString())
                .apply { envelope.getDeviceId()?.let { header("X-Device-Id", it) } }
                .header("X-Platform", envelope.getPlatformName())
                .apply { envelope.getAppVersion()?.let { header("X-App-Version", it) } }
                .header("X-OS-Version", envelope.getOsVersion())
                // TODO(edge): placeholder — edge worker will set these from client IP
                .header("X-Resolved-Country", "US")
                .header("X-Resolved-Market-Cohort", "MARKET_UNKNOWN")
                .apply { authorizationToken?.let { header("Authorization", "Bearer $it") } }
                .post(envelope.buildJsonBody().toRequestBody(JSON_MEDIA_TYPE))
                .build()
        } else {
            // Standard GET with bracket-notation query params + user params.
            val combined = if (userQuery.isEmpty()) envelopeQuery else "$userQuery&$envelopeQuery"
            val separator = if (path.contains("?")) "&" else "?"
            val url = "$baseUrl$path${separator}$combined"
            Log.d(TAG, "Fetching screen: $url")
            Request.Builder()
                .url(url)
                .header("X-Trace-Id", traceId)
                .header("X-Request-Id", UUID.randomUUID().toString())
                .apply { envelope.getDeviceId()?.let { header("X-Device-Id", it) } }
                .header("X-Platform", envelope.getPlatformName())
                .apply { envelope.getAppVersion()?.let { header("X-App-Version", it) } }
                .header("X-OS-Version", envelope.getOsVersion())
                // TODO(edge): placeholder — edge worker will set these from client IP
                .header("X-Resolved-Country", "US")
                .header("X-Resolved-Market-Cohort", "MARKET_UNKNOWN")
                .apply { authorizationToken?.let { header("Authorization", "Bearer $it") } }
                .build()
        }

        val response = httpClient.newCall(request).execute()
        if (!response.isSuccessful) {
            throw SduiException("Failed to fetch screen: ${response.code}")
        }

        // Detect server-signaled schema version mismatch (upgrade-required)
        val versionMismatch = response.header("X-Schema-Version-Mismatch")
        if (versionMismatch == "upgrade-required") {
            Log.w(TAG, "Server signaled schema version mismatch: upgrade-required")
            throw SchemaVersionMismatchException(
                "Client schema version is no longer supported. Please update the app."
            )
        }

        val body = response.body?.string()
            ?: throw SduiException("Empty response body")

        try {
            mapper.readValue(body, SduiModels::class.java)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse screen response", e)
            throw SduiException("Failed to parse screen response: ${e.message}")
        }
    }

    /**
     * Percent-encode user params with the same RFC-3986 rule the envelope
     * uses, sorted by key so the encoded URL is stable across calls (matters
     * for parity tests and CDN cache keys).
     */
    private fun encodeUserParams(params: Map<String, String>): String {
        if (params.isEmpty()) return ""
        return params.entries
            .filter { it.value.isNotEmpty() }
            .sortedBy { it.key }
            .joinToString("&") { (k, v) ->
                "${RequestEnvelopeBuilder.percentEncode(k)}=${RequestEnvelopeBuilder.percentEncode(v)}"
            }
    }

    /**
     * Fetch raw JSON from any URL (for direct data polling).
     *
     * @param url The full URL to fetch from
     * @param dataPath Optional JSONPath-like path to extract data (e.g., "game" or "sections[0].data")
     * @param traceId Optional trace ID to send via X-Trace-Id header for log correlation
     */
    suspend fun fetchRawJson(url: String, dataPath: String? = null, traceId: String? = null): Map<String, Any> = withContext(Dispatchers.IO) {
        Log.d(TAG, "Fetching raw JSON from: $url${traceId?.let { " [trace=$it]" } ?: ""}")

        val requestBuilder = Request.Builder().url(url)
        if (traceId != null) {
            requestBuilder.header("X-Trace-Id", traceId)
        }
        val request = requestBuilder.build()

        val response = httpClient.newCall(request).execute()

        if (!response.isSuccessful) {
            throw SduiException("Failed to fetch from $url: ${response.code}")
        }

        val body = response.body?.string()
            ?: throw SduiException("Empty response body from $url")

        @Suppress("UNCHECKED_CAST")
        val json = mapper.readValue(body, Map::class.java) as Map<String, Any>

        // Extract nested path if specified (simple dot notation)
        if (dataPath.isNullOrEmpty()) {
            return@withContext json
        }

        var current: Any = json
        for (segment in dataPath.split(".")) {
            current = when (current) {
                is Map<*, *> -> current[segment] ?: throw SduiException("Path '$segment' not found in response")
                else -> throw SduiException("Cannot navigate path '$segment' - current value is not a map")
            }
        }

        @Suppress("UNCHECKED_CAST")
        when (current) {
            is Map<*, *> -> current as Map<String, Any>
            else -> mapOf("data" to current)
        }
    }
}

/**
 * Exception for SDUI-related errors.
 */
class SduiException(message: String, cause: Throwable? = null) : Exception(message, cause)

/**
 * Thrown when the server signals that the client's schema version is below the
 * minimum supported version via `X-Schema-Version-Mismatch: upgrade-required`.
 * The ViewModel should catch this and display an app-update prompt.
 */
class SchemaVersionMismatchException(message: String) : Exception(message)

File: ios/Sources/SduiCore/Network/SduiRepository.swift
import Foundation
import os

private let logger = Logger(subsystem: "com.nba.sdui", category: "SduiRepository")

/// Single entry point for all SDUI network requests — one generic fetch
/// method, no dedicated `getGameDetail()` helpers.
///
/// Transport contract (ADR-003 / plan-request-transport):
/// - Bracket-notation GET query params below 8192 chars, otherwise POST with
///   the identical envelope as a JSON body.
/// - Only `Authorization` and `X-Trace-Id` headers. Deprecated `X-Platform`
///   and `X-Schema-Version` headers are no longer sent.
final class SduiRepository {
    private let config: SduiConfig
    private let session: URLSession
    private let envelopeProvider: @Sendable () -> RequestEnvelope
    private let fetchStateLock = NSLock()
    private var activeScreenFetch: Task<SduiModels, Error>?
    /// Per-request `X-Trace-Id` of the current in-flight screen fetch. Reused
    /// as the active-fetch identity so the deferred cleanup only nils the slot
    /// when no newer fetch has overwritten it (Task is a struct, so identity
    /// comparison via `===` is unavailable).
    private var activeScreenFetchTraceID: String?

    deinit {
        fetchStateLock.withLock { activeScreenFetch?.cancel() }
    }

    init(
        config: SduiConfig,
        session: URLSession = .shared,
        envelopeProvider: @Sendable @escaping () -> RequestEnvelope = { RequestEnvelope() }
    ) {
        self.config = config
        self.session = session
        self.envelopeProvider = envelopeProvider
    }

    /// Fetch a screen from any SDUI endpoint. The endpoint is always provided
    /// by caller/server — never hardcoded.
    ///
    /// - Parameters:
    ///   - endpoint: server-relative path (e.g. `/sdui/scoreboard`).
    ///   - userParams: optional user-supplied filter params (e.g. Form submit
    ///     bindings like `season=2025-26`). Always travel in the URL query
    ///     string regardless of GET vs POST so the server reads them
    ///     uniformly. They participate in the GET/POST length decision.
    ///   - traceID: optional trace ID to reuse from a parent fetch (e.g.
    ///     parameterized refresh inheriting its screen's trace). Falls back
    ///     to the config's provider when absent.
    func fetchScreen(
        endpoint: String,
        userParams: [String: String] = [:],
        traceID: String? = nil
    ) async throws -> SduiModels {
        let envelope = envelopeProvider()

        let resolvedTraceID = traceID ?? config.traceIDProvider()
        let request = try buildRequest(
            endpoint: endpoint,
            envelope: envelope,
            userParams: userParams,
            traceID: resolvedTraceID
        )
        logger.debug("Fetching screen \(request.httpMethod ?? "GET"): \(request.url?.absoluteString ?? endpoint) [trace=\(resolvedTraceID, privacy: .public)]")

        fetchStateLock.withLock { activeScreenFetch?.cancel() }

        let httpSession = self.session
        let task: Task<SduiModels, Error> = Task {
            let (data, response) = try await httpSession.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse else {
                throw SduiError.invalidResponse
            }

            guard (200...299).contains(httpResponse.statusCode) else {
                logger.error("HTTP \(httpResponse.statusCode) for \(endpoint)")
                throw SduiError.httpError(statusCode: httpResponse.statusCode)
            }

            // Detect server-signaled schema version mismatch
            if httpResponse.value(forHTTPHeaderField: "X-Schema-Version-Mismatch") == "upgrade-required" {
                logger.warning("Server signaled schema version mismatch: upgrade-required")
                throw SduiError.upgradeRequired
            }

            do {
                let screen = try SduiModels(data: data)
                logger.debug("Fetched screen '\(screen.id)' with \(screen.sections.count) sections")
                return screen
            } catch {
                let detail = Self.describeDecodingError(error)
                logger.error("Failed to decode screen from \(endpoint, privacy: .public): \(detail, privacy: .public)")
                throw SduiError.decodingFailed(detail)
            }
        }
        fetchStateLock.withLock {
            activeScreenFetch = task
            activeScreenFetchTraceID = resolvedTraceID
        }
        defer {
            fetchStateLock.withLock {
                if activeScreenFetchTraceID == resolvedTraceID {
                    activeScreenFetch = nil
                    activeScreenFetchTraceID = nil
                }
            }
        }
        return try await task.value
    }

    /// Expand a `DecodingError` into `keyNotFound/typeMismatch/valueNotFound/dataCorrupted`
    /// with the failing JSON path and the underlying reason. `localizedDescription`
    /// hides that detail, which makes on-device decode failures unreadable.
    private static func describeDecodingError(_ error: Error) -> String {
        guard let decodingError = error as? DecodingError else {
            return error.localizedDescription
        }
        let path: (DecodingError.Context) -> String = { ctx in
            ctx.codingPath.map { $0.stringValue }.joined(separator: ".")
        }
        switch decodingError {
        case let .keyNotFound(key, ctx):
            return "keyNotFound '\(key.stringValue)' at '\(path(ctx))' — \(ctx.debugDescription)"
        case let .typeMismatch(type, ctx):
            return "typeMismatch expected=\(type) at '\(path(ctx))' — \(ctx.debugDescription)"
        case let .valueNotFound(type, ctx):
            return "valueNotFound \(type) at '\(path(ctx))' — \(ctx.debugDescription)"
        case let .dataCorrupted(ctx):
            return "dataCorrupted at '\(path(ctx))' — \(ctx.debugDescription)"
        @unknown default:
            return String(describing: decodingError)
        }
    }

    /// Fetch raw JSON from a direct URL (for poll refresh with `dataPath` extraction).
    /// - Parameter traceID: Optional trace ID to reuse from the parent screen fetch.
    ///   Falls back to generating a fresh ID if nil.
    func fetchRawJson(url: URL, traceID: String? = nil) async throws -> Any {
        var request = URLRequest(url: url)
        attachAuthHeaders(&request, traceID: traceID ?? config.traceIDProvider())

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw SduiError.invalidResponse
        }

        return try JSONSerialization.jsonObject(with: data)
    }

    // MARK: - Request assembly

    /// Build a request for the canonical SDUI transport. The envelope owns the
    /// GET-vs-POST decision (length-based via `exceedsGetThreshold`); user
    /// params always travel in the URL query string regardless of method, so
    /// the server reads them through the same `@RequestParam` path on either
    /// side. Both halves of the query string use RFC-3986 percent-encoding so
    /// values with `&`, `=`, spaces, or non-ASCII bytes survive the round-trip
    /// intact.
    private func buildRequest(
        endpoint: String,
        envelope: RequestEnvelope,
        userParams: [String: String],
        traceID: String
    ) throws -> URLRequest {
        let path = endpoint.hasPrefix("/") ? endpoint : "/\(endpoint)"
        let baseWithPath = config.baseURL.appendingPathComponent(
            String(path.dropFirst())
        )

        let userQuery = Self.encodeUserParams(userParams)
        // Threshold includes both halves so large userParams trigger POST too.
        let envelopeQuery = envelope.buildQueryString()
        let combinedLength = envelopeQuery.count + (userQuery.isEmpty ? 0 : userQuery.count + 1)

        if combinedLength > RequestEnvelope.maxQueryLength {
            let postURL: URL
            if userQuery.isEmpty {
                postURL = baseWithPath
            } else {
                guard let url = URL(string: baseWithPath.absoluteString + "?" + userQuery) else {
                    throw SduiError.invalidURL(endpoint)
                }
                postURL = url
            }
            var request = URLRequest(url: postURL)
            request.httpMethod = "POST"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try envelope.jsonBody()
            attachAuthHeaders(&request, traceID: traceID, envelope: envelope)
            return request
        }

        let combined = userQuery.isEmpty ? envelopeQuery : userQuery + "&" + envelopeQuery
        let urlString = baseWithPath.absoluteString + "?" + combined
        guard let url = URL(string: urlString) else {
            throw SduiError.invalidURL(endpoint)
        }
        var request = URLRequest(url: url)
        attachAuthHeaders(&request, traceID: traceID, envelope: envelope)
        return request
    }

    /// Percent-encode user params with the same RFC-3986 rule the envelope
    /// uses, preserving deterministic order so the encoded URL is stable
    /// across calls (matters for parity tests and CDN cache keys).
    private static func encodeUserParams(_ params: [String: String]) -> String {
        guard !params.isEmpty else { return "" }
        return params
            .filter { !$0.value.isEmpty }
            .sorted(by: { $0.key < $1.key })
            .map { "\(RequestEnvelope.percentEncode($0.key))=\(RequestEnvelope.percentEncode($0.value))" }
            .joined(separator: "&")
    }

    private func attachAuthHeaders(_ request: inout URLRequest, traceID: String, envelope: RequestEnvelope? = nil) {
        if let token = config.authorizationToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        request.setValue(traceID, forHTTPHeaderField: "X-Trace-Id")
        request.setValue(UUID().uuidString, forHTTPHeaderField: "X-Request-Id")
        if let deviceID = envelope?.deviceID {
            request.setValue(deviceID, forHTTPHeaderField: "X-Device-Id")
        }
        if let envelope {
            request.setValue(envelope.platformName, forHTTPHeaderField: "X-Platform")
            if let appVersion = envelope.appVersion {
                request.setValue(appVersion, forHTTPHeaderField: "X-App-Version")
            }
            request.setValue(envelope.osVersion, forHTTPHeaderField: "X-OS-Version")
        }
        // TODO(edge): placeholder — edge worker will set these from client IP
        request.setValue("US", forHTTPHeaderField: "X-Resolved-Country")
        request.setValue("MARKET_UNKNOWN", forHTTPHeaderField: "X-Resolved-Market-Cohort")
    }
}

enum SduiError: LocalizedError, Equatable {
    case invalidURL(String)
    case invalidResponse
    case httpError(statusCode: Int)
    case decodingFailed(String)
    case upgradeRequired

    var errorDescription: String? {
        switch self {
        case .invalidURL(let path): return "Invalid URL for endpoint: \(path)"
        case .invalidResponse: return "Invalid HTTP response"
        case .httpError(let code): return "HTTP error: \(code)"
        case .decodingFailed(let message): return "Decoding failed: \(message)"
        case .upgradeRequired: return "This version of the app is no longer supported. Please update to continue."
        }
    }
}
